File: lib/rag/ingest.ts
import { nanoid } from "nanoid";
import { chunkPages } from "./chunk";
import { parseFile } from "./parse";
import { embedDocuments } from "./embeddings";
import {
  addChunks,
  registerDocument,
  type ChunkRow,
  type DocumentMeta,
} from "./vectorstore";

/** Full ingestion pipeline: parse -> chunk -> embed -> store. */
export async function ingestFile(
  buffer: Buffer,
  filename: string,
  mime: string,
): Promise<DocumentMeta> {
  const pages = await parseFile(buffer, filename, mime);
  const chunks = chunkPages(pages);
  if (chunks.length === 0) {
    throw new Error("Tidak ada teks yang bisa diekstrak dari dokumen ini.");
  }

  const vectors = await embedDocuments(chunks.map((c) => c.text));
  const docId = nanoid();

  const rows: ChunkRow[] = chunks.map((c, i) => ({
    id: `${docId}:${i}`,
    docId,
    docName: filename,
    page: c.page,
    chunkIndex: i,
    text: c.text,
    vector: vectors[i],
  }));

  await addChunks(rows);

  const meta: DocumentMeta = {
    id: docId,
    name: filename,
    size: buffer.length,
    chunks: rows.length,
    createdAt: new Date().toISOString(),
  };
  await registerDocument(meta);
  return meta;
}

File: backend/app/rag/ingest.py
"""Ingestion pipeline: parse -> chunk -> embed -> store + keep the original file."""

import uuid
from datetime import datetime, timezone

from .chunk import chunk_pages
from .embeddings import embed_documents
from .faiss_store import ChunkRow, get_store
from .files import ext_of, save_original
from .parse import parse_file


async def ingest_file(buffer: bytes, filename: str, mime: str) -> dict:
    pages = parse_file(buffer, filename, mime)
    chunks = chunk_pages(pages)
    if not chunks:
        raise ValueError("No extractable text found in this document.")

    vectors = await embed_documents([c["text"] for c in chunks])
    doc_id = uuid.uuid4().hex

    rows = [
        ChunkRow(
            doc_id=doc_id,
            doc_name=filename,
            page=c["page"],
            chunk_index=i,
            text=c["text"],
            vector=vectors[i],
        )
        for i, c in enumerate(chunks)
    ]

    store = get_store()
    store.add_chunks(rows)

    ext = ext_of(filename)
    save_original(doc_id, ext, buffer)

    meta = {
        "id": doc_id,
        "name": filename,
        "size": len(buffer),
        "chunks": len(rows),
        "createdAt": datetime.now(timezone.utc).isoformat(),
        "ext": ext,
        "mime": mime or None,
    }
    store.register_document(meta)
    return meta
